src/test/java/jenkins/branch/naming/ProjectNamingStrategyTest.java

```diff
@@ -0,0 +1,162 @@
+package jenkins.branch.naming;
+
+import com.sun.net.httpserver.HttpServer;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.logging.Logger;
+import jenkins.branch.BranchSource;
+import jenkins.branch.MultiBranchProjectDisplayNamingStrategy;
+import jenkins.branch.MultiBranchProjectDisplayNamingTrait;
+import jenkins.branch.naming.githubapimock.MockGithubInfo;
+import jenkins.branch.naming.githubapimock.MockGithubJenkinsfileContent;
+import jenkins.branch.naming.githubapimock.MockGithubOrg;
+import jenkins.branch.naming.githubapimock.MockGithubRateLimit;
+import jenkins.branch.naming.githubapimock.MockGithubRepository;
+import jenkins.branch.naming.githubapimock.MockGithubUser;
+import jenkins.branch.naming.githubapimock.MockPullRequest;
+import jenkins.branch.naming.githubapimock.ResourceHandler;
+import jenkins.model.Jenkins;
+import jenkins.scm.api.SCMSource;
+import org.jenkinsci.plugins.github_branch_source.GitHubSCMSource;
+import org.jenkinsci.plugins.github_branch_source.OriginPullRequestDiscoveryTrait;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;
+import org.junit.*;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * This class tests that project naming strategies (see {@link MultiBranchProjectDisplayNamingStrategy}) applied to
+ * multi-branch jobs via {@link MultiBranchProjectDisplayNamingTrait} are used to set the display name of change
+ * request jobs inside the multi-branch project.
+ *
+ * It does this by using the github-branch-source plugin and stubbing the GitHub REST API to create a fake change
+ * request. Then it checks the job's display name against what
+ * {@link MultiBranchProjectDisplayNamingStrategy#generateName(String, String)} generates for that change request.
+ *
+ * Another option for testing this (that wouldn't have us rely on another plugin) would be to implement our own
+ * fake {@link SCMSource} but that's probably even more complex. Since the github-branch-source plugin is one of the
+ * most used (and fingers-crossed, stable) in the Jenkins ecosystem, and mocking a REST API is trivial, the tradeoff is
+ * acceptable.
+ */
+public class ProjectNamingStrategyTest {
+    private static final Logger LOGGER = Logger.getLogger(ProjectNamingStrategyTest.class.getName());
+
+    @ClassRule
+    public static final JenkinsRule r = new JenkinsRule();
+
+    public static String stubBaseUrl ;
+    public static HttpServer githubApiStub;
+
+    @BeforeClass
+    public static void setupGithubApiStub() throws Exception {
+        final int stubPort = findAvailablePort();
+        stubBaseUrl = String.format("http://localhost:%s", stubPort);
+        githubApiStub = HttpServer.create(new InetSocketAddress(stubPort), 0);
```

Comment:
See comment below on `findAvailablePort()`  
Additionally this binds to every interface which seems overkill and one jdk exploit / firewall misconfiguration away from an accident.


Untested but something like the following would be preferable.

```java
        githubApiStub = HttpServer.create(new InetSocketAddress(netAddress.getLoopbackAddress(), 0), 0);

       // all the other code as the port will not be available until we `start` the server

        // obtain the actual bound address.
        final InetSocketAddress boundSocket = githubApiStub.getAddress();
        final int port = boundSocket.getPort();

        final InetAddress addr = boundSocket.getAddress();
        // literal IPv6 addresses need special encoding
        final String address;
        if (addr instanceOf Inet6Address) {
            address = "[" + addr.getHostAddress()) + "]";
        } else {
            address = addr.getHostAddress());
       }
        stubBaseUrl = String.format("http://%s:%d", address, port);
```

Reply:
I'll try this :+1:

Reply:
Tested OK, and added.